src/foam/dao/pg/PostgresDAO.java

```diff
@@ -0,0 +1,326 @@
+package foam.dao.pg;
+
+import foam.core.ClassInfo;
+import foam.core.FObject;
+import foam.core.PropertyInfo;
+import foam.core.X;
+import foam.dao.AbstractDAO;
+import foam.dao.ListSink;
+import foam.dao.SQLType;
+import foam.dao.Sink;
+import foam.mlang.order.Comparator;
+import foam.mlang.predicate.Predicate;
+
+import java.sql.*;
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class PostgresDAO
+    extends AbstractDAO
+{
+  protected ConnectionPool connectionPool = new ConnectionPool();
+  protected ThreadLocal<StringBuilder> sb = new ThreadLocal<StringBuilder>() {
+    @Override
+    protected StringBuilder initialValue() {
+      return new StringBuilder();
+    }
+
+    @Override
+    public StringBuilder get() {
+      StringBuilder b = super.get();
+      b.setLength(0);
+      return b;
+    }
+  };
+
+  protected final String table;
+
+  public PostgresDAO(ClassInfo of, String host, String port, String dbName, String username, String password) throws SQLException {
+    setOf(of);
+
+    if ( dbName == null || username == null ) {
+      throw new IllegalArgumentException("Illegal arguments");
+    }
+
+    host = (host != null) ? host : "localhost";
+    port = (port != null) ? port : "5432";
+
+    // set up connection pool
+    connectionPool.setup(host, port, dbName, username, password);
+
+    // load columns and sql types
+    table = of.getObjClass().getSimpleName().toLowerCase();
+    createTable(of);
+  }
+
+  public Sink select_(X x, Sink sink, long skip, long limit, Comparator order, Predicate predicate) {
+    if ( sink == null ) {
+      sink = new ListSink();
+    }
+
+    try {
+      Connection c = connectionPool.getConnection();
+      PreparedStatement stmt = null;
+
+      // select all if predicate is null, else use predicate
+      if ( predicate == null ) {
+        StringBuilder builder = sb.get()
+            .append("select * from ")
+            .append(table);
+        stmt = c.prepareStatement(builder.toString());
+      } else {
+        stmt = c.prepareStatement(predicate.createStatement(table));
+      }
+
+      ResultSet resultSet = stmt.executeQuery();
+      while ( resultSet.next() ) {
+        sink.put(createFObject(resultSet), null);
+      }
+
+      resultSet.close();
```

Comment:
More care needed for exception handling and closing resources. If you get an exception you won't close your resultSet, stmt and connection, causing your connections to be drained from the thread-pool and blocking your whole server.

Reply:
Thanks. I modified all the closing of resources.